api/server/error_response.go

```diff
@@ -46,7 +50,11 @@ func HandleErrorResponse(ctx context.Context, w http.ResponseWriter, err error)
 		}
 		statuscode = e.Code()
 	} else {
-		log.WithError(err).WithFields(logrus.Fields{"stack": string(debug.Stack())}).Error("internal server error")
+		if le, ok := err.(models.LoggableError); ok {
+			log.WithError(err).WithFields(logrus.Fields{"stack": string(debug.Stack())}).Error(le.Message())
+		} else {
+			log.WithError(err).WithFields(logrus.Fields{"stack": string(debug.Stack())}).Error("internal server error")
+		}
 		statuscode = http.StatusInternalServerError
 		err = ErrInternalServerError
 	}
```

Comment:
after discussion and thinking on this, I've made the following table of assumptions, we have 3 (not distinct) kinds of errors presented here:

a) a `models.APIError`, whose error message should be both logged as well as output & presents statusCode
b) a `models.LoggableError`, whose error message should be logged but not output & 500 (in the case its <500, it's not logged)
c) a plain-ole-err, whose error message should be logged but not output & 500

in this respect, I don't find much of a difference between b & c, aside from an additional message that we are adding to errors (could this be encoded in the error? why or why not?), I am not sure of the value exactly there and can't imagine it's anything more descriptive than 'internal server error' without us wanting to use the more heavy-handed models.APIError to output something useful. I do see the fact that when using a models.APIError, it is both logged and output, however passing any non-models.APIError seems to give the value of logging the error but not outputting it. this begs the question, are there models.APIErrors where we want to encode a additional information to the user but not output that error? in any event, after drawing these out, i'm less sure of the utility provided here (all this just to change the log message line when we just need to pass a non-models.APIError?) and am working off the 'keep things as simple as possible' principle. could we clarify the value proposition here? what am I missing :) ?

Reply:
My intent was to allow for cases where we want the HTTP request to return an error code other than 500 and we want to log some specifics about what happened (but not include all of those specifics in the HTTP response since it's nothing that's actionable by the user). The first use case I had in mind is a case where we currently return a 503 -- so I'd like to continue to return a 503, but log some additional information associated with the request.

Also, I hadn't noticed that we'll only log the message if the status is >=500. Is there any reason that couldn't be >= 400 instead?

Reply:
> The first use case I had in mind is a case where we currently return a 503 -- so I'd like to continue to return a 503, but log some additional information associated with the request.

gotcha, from above I wasn't clear on how this would be different than c) - or passing an error that is _not_ of models.APIError, however, it seems like the case of primary interest _is_ a models.APIError and we want to decorate it with some additional information (to log, not to the user). I think that puts me on the same page? please correct if not. in any event, I'm not sure the message section is the best spot for this, it seems like we want the 'real error' and not the models.APIError masking it?

in the past I've seen this done this way https://github.com/iron-io/runner/blob/master/common/errors.go#L24 - I can't admit to being a fan (the bool is unnecessary, for one, either implement it or don't), but maybe it gets the idea juices flowing. the models.APIError thing does suppress errors, especially in the 5xx case, so I kind of buy this 2 different kinds of error thing.

for our 5xx errors, would something like this make sense or am I missing things?

```go
// LogErr attaches an additional error onto a models.APIError
type LoggableErr interface {
  models.APIError
  LogError() error
}

type loggableErr struct {
  models.APIError
  root error
}

func (le loggableErr) LogError() error { return le.root }

func NewLoggableAPIError(apiErr, rootErr error) LoggableErr {
  return &loggableErr{
    APIError: apiErr,
    root: rootErr,
  }
}

// where we're interested in handleErrResponse...
log := common.Logger(ctx)
if le, ok := (models.LoggableError); ok {
  log = log.WithField("root_error": le.LogError())
}

// leave the rest as is with the models.APIError dance ?...
```

this would prevent 'error suppression' from models.APIErrors (we could get at the root error), we wouldn't need to use the message field (confusing) with eg `err=xyz root_err=abc`, and most errors could use the old error trap for 503 (`c)` from above). I'm not sure it's the best solution or the one you're going for, if I'm still missing the point please let me know, and worst case maybe that offers some ideas.

> Also, I hadn't noticed that we'll only log the message if the status is >=500. Is there any reason that couldn't be >= 400 instead?

noise would be the primary argument against doing so, we could push muting things to the logging backend though (ie every `not found` gets logged, and that's on the user / not very informative to us usually). for OSS, it would make some sense though to log every >= 400.

Reply:
the inimitable dave cheney has done it this way https://github.com/pkg/errors which I think turned out pretty slick, adopting something in this vein would make sense (previous efforts to adopt the pkg/errors package wholeheartedly have not worked out well, but for our api error suppression, something along these lines makes sense)

Reply:
Nice, I like this better. Implemented something along the lines of what you suggested.